assets/app/scripts/services/applicationGenerator.js

```diff
@@ -0,0 +1,270 @@
+"use strict";
+
+angular.module("openshiftConsole")
+
+  .service("ApplicationGenerator", function(){
+    var osApiVersion = "v1beta1";
+    var k8sApiVersion = "v1beta3";
+    
+    var service = {};
+    var scope = service;
+    
+    scope._generateSecret_ = function(){
+        //http://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
+        function s4() {
+          return Math.floor((1 + Math.random()) * 0x10000)
+              .toString(16)
+              .substring(1);
+          }
+        return s4()+s4()+s4()+s4();
+      };
+
+    /**
+    * Find the 'first' port of exposed ports.
+    * @param            ports  list of ports (e.g {containerPort: 80, protocol: "tcp"})
+    * @return {integer} The port/protocol pair of the lowest conainer port
+    */
+    scope._getFirstPort_ = function(ports){
+      var first = "None";
+      ports.forEach(function(port){
+        if(first === "None"){
+            first = port;
+          }else{
+            if(port.containerPort < first.containerPort){
+              first = port;
+            }
+          }
+        }
+      );
+      return first;
+    };
+    
+    /**
+     * Generate resource definitions to support the given input
+     * @param {type} input
+     * @returns Hash of resource definitions
+     */
+    scope.generate = function(input){
+      //map ports to k8e structure
+      var ports = [];
+      angular.forEach(input.image.dockerImageMetadata.ContainerConfig.ExposedPorts, function(value, key){
+        var parts = key.split("/");
+        if(parts.length === 1){
+          parts.push("tcp");
+        }
+        ports.push(
+          {
+            containerPort: parseInt(parts[0]), 
+            name: input.name + "-" + parts[1] + "-" + parts[0],
+            protocol: parts[1]
+          });
+      });
+
+      //augment labels
+      input.labels.name = input.name;
+      input.labels.generatedby = "OpenShiftWebConsole";
+      
+      var imageSpec;
+      if(input.buildConfig.sourceUrl !== null){
+        imageSpec = {
+          name: input.name, 
+          tag: "latest",
+          toString: function(){
+            return this.name + ":" + this.tag;
```

Comment:
Isn't the use of 'this' more correct since 'this' will refer to the hash 
that is imageSpec?

On 04/08/2015 03:24 PM, Jessica Forrester wrote:
> In assets/app/scripts/services/applicationGenerator.js
> <https://github.com/openshift/origin/pull/1221#discussion_r28003944>:
>
>> +            name: input.name + "-" + parts[1] + "-" + parts[0],
>> +            protocol: parts[1]
>> +          });
>> +      });
>> +
>> +      //augment labels
>> +      input.labels.name = input.name;
>> +      input.labels.generatedby = "OpenShiftWebConsole";
>> +
>> +      var imageSpec;
>> +      if(input.buildConfig.sourceUrl !== null){
>> +        imageSpec = {
>> +          name: input.name,
>> +          tag: "latest",
>> +          toString: function(){
>> +            return this.name + ":" + this.tag;
>
> above you reference input.name, here you reference this.name. Is there a
> difference? If not lets be consistent, one way or the other.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1221/files#r28003944>.
>

-- 
Jeff Cantrill
Senior Software Engineer, Red Hat Engineering
Red Hat
Office: 703-748-4420 | 866-546-8970 ext. 8162420
jcantril@redhat.com
http://www.redhat.com